src/core/syscalls/business_logic_syscall_handler.rs

```diff
@@ -161,28 +180,82 @@ impl SyscallHandler for BusinessLogicSyscallHandler {
-    ) -> Result<(), SyscallHandlerError> {
-        let request = match self._read_and_validate_syscall_request("emit_event", vm, syscall_ptr) {
-            Ok(SyscallRequest::EmitEvent(emit_event_struct)) => emit_event_struct,
-            _ => return Err(SyscallHandlerError::InvalidSyscallReadRequest),
-        };
-
-        let keys_len = request.keys_len;
-        let data_len = request.data_len;
+    ) -> Result<Vec<u64>, SyscallHandlerError> {
+        // Parse request and prepare the call.
+        let request =
+            match self._read_and_validate_syscall_request(syscall_name, vm, syscall_ptr)? {
+                SyscallRequest::CallContract(request) => request,
+                _ => return Err(SyscallHandlerError::ExpectedCallContract),
+            };
+
+        let calldata = get_integer_range(vm, &request.calldata, request.calldata_size)?;
+
+        let mut code_address = None;
+        let mut class_hash = None;
+
+        match syscall_name {
+            "call_contract" => {
+                code_address = Some(request.contract_address);
+                let contract_address = code_address;
+                let caller_address = self.contract_address;
+                let entry_point_type = EntryPointType::External;
+                let call_type = CallType::Call;
+            }
+            "delegate_call" => {
+                code_address = Some(request.contract_address);
+                let contract_address = self.contract_address;
+                let caller_address = self.caller_address;
+                let entry_point_type = EntryPointType::External;
+                let call_type = CallType::Delegate;
+            }
+            "delegate_l1_handler" => {
+                code_address = Some(request.contract_address);
+                let contract_address = self.contract_address;
+                let caller_address = self.caller_address;
+                let entry_point_type = EntryPointType::L1Handler;
+                let call_type = CallType::Delegate;
+            }
+            "library_call" => {
+                class_hash = Some(request.class_hash.to_be_bytes());
+                let contract_address = self.contract_address;
+                let caller_address = self.caller_address;
+                let entry_point_type = EntryPointType::External;
+                let call_type = CallType::Delegate;
+            }
+            "library_call_l1_handler" => {
+                class_hash = Some(request.class_hash.to_be_bytes());
+                let contract_address = self.contract_address;
+                let caller_address = self.caller_address;
+                let entry_point_type = EntryPointType::L1Handler;
+                let call_type = CallType::Delegate;
+            }
+            _ => {
+                return Err(SyscallHandlerError::UnknownSyscall(
+                    syscall_name.to_string(),
+                ))
+            } // TODO add some message to this error.
```

Comment:
I think we can remove this TODO
I am right?

Reply:
Oh yes, you are right. My bad. Fixing it rn.